### PR TITLE
Migrate module configurations in RecoLocalMuon to use automatically generated cfi default configurations

### PR DESCRIPTION
**Pull request by jeongeun**
#### PR description: 
Optimization of the python configurations: Improve maintainability by cleaning up the duplicated and cloning from the default/reference configurations.
 
In this PR, 1 file changed.  (The previous PRs were  [PR#33207](https://github.com/cms-sw/cmssw/pull/33207), [PR#33307](https://github.com/cms-sw/cmssw/pull/33307), [PR#33352](https://github.com/cms-sw/cmssw/pull/33352) )

- commit 1 : 
Replace explicit configuration with a reference from cfipython/. (migrating EDProducer("type", .. -> typeDefaylt.clone() )
Remove the type specifications already presented in cfipython/fillDescriptions reference for improved syntax safety.

- commit 2 : 
Remove the duplicated parameters that are exactly the same value in cfipython reference.

#### PR validation:
Event Content comparison check was also done and there is no change with these updates.
Tested in CMSSW_11_3_X, the basic test all passed in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html).

### CODE REVIEW

File: RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cfi.py [hunk 1]
```diff
@@ -4,73 +4,25 @@
...
-    #    How to find SCA peak time?
-    #                              
-    UseAverageTime = cms.bool(False),
-    UseParabolaFit = cms.bool(False),
-    UseFivePoleFit = cms.bool(True),                       
-    #
-    #    Parameters for wire hits
-    CSCWireClusterDeltaT = cms.int32(1),
-    #
+import RecoLocalMuon.CSCRecHitD.configWireTimeWindow_cfi as _mod
```
Comment (perrotta) [inline RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cfi.py:7]: @ptcox this PR aims at reusing the default configuration as defined in the `fillDescription` method of `CSCRecHitDProducer`, see item 6 in https://twiki.cern.ch/twiki/bin/view/CMS/OpenRecoTasks#Review_and_optimization_of_t_AN1).

In this case, by removing the un-modified parameters one also removes the comment lines inserted, and possibly an easy modification of the configuration. Therefore, I'd be tempted to ask to revert the second part of this PR and limit it to the part in which the configuration is cloned from the "default" one, and parameters are set modifying the original ones, i.e. what coded in the [first commit](https://github.com/cms-sw/cmssw/pull/33354/commits/35cc9d02c3d606fd5465a7a66db56df7d54d6795) of this PR

Before asking @jeongeun to revert the second commit: do you have any suggestion or comment, Tim? Would you stick on the "thinner" configuration resulting from the current version of this PR, instead?

**Comment by ptcox** [comment]
Hi Andrea,

Thanks for drawing my attention to this. I agree with you that it would 
be MUCH MUCH BETTER to retain the comments (and the structure).

(This is a big problem with changing things to use fillDescription. What 
is a default NOW might not have been a default when the code was first 
developed. And because the organization shown via comments gets lost, 
one can easily end up with a complete lack of understanding of which 
config parameters control what.)

Regards, Tim

perrotta wrote on 4/7/21 17:27:

**Comment by perrotta** [comment]
Thank you @ptcox

@jeongeun could you then please revert your second commit, as we anticipated yesterday during our chat, and re-implement the call to the external `cscRecHitDParameters`?

**Comment by perrotta** [comment]
Kind remind @jeongeun

> @jeongeun could you then please revert your second commit, as we anticipated yesterday during our chat, and re-implement the call to the external `cscRecHitDParameters`?

File: RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cfi.py [hunk 2]
```diff
@@ -4,73 +4,75 @@
...
+
+csc2DRecHits = _mod.configWireTimeWindow.clone(
     #
     #    Parameters for coordinate and uncertainty calculations
     #    Data and MC parameters are (still) different
     #    Needs tuning
     #
-    cscRecHitDParameters,
+    #cscRecHitDParameters, # drop this due to identical values in cfipython _mod
```
Comment (perrotta) [inline RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cfi.py:15]: For the same reason as we put back the other parameters in the explicit config, I would uncomment this line (and remove the comment): developers can then modify `cscRecHitDParameters` in `RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cff.py` without bothering about uncommenting the line here

**Comment by perrotta** [comment]
please test

**Comment by jeongeun** [comment]
@perrotta 
I'm not sure how to make it works. ValueError has occurred as below, after uncomment "cscRecHitDParameters,".
I checked the cscRecHitDParameters ([RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cff.py](https://github.com/cms-sw/cmssw/blob/master/RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cff.py)) and it exactly same as the [cfipy](https://cmssdt.cern.ch/lxr/source/cfipython/RecoLocalMuon/CSCRecHitD/configWireTimeWindow_cfi.py).

Is that okay to comment out this "cscRecHitDParameters," ? please advice on this error.
Thanks.

```
ValueError: Duplicate insert of member XTasymmetry_ME1b
The original parameters are:
cms.EDProducer("CSCRecHitDProducer",
    ConstSyst_ME12 = cms.double(0.02),
    ConstSyst_ME13 = cms.double(0.03),
    ConstSyst_ME1a = cms.double(0.01),
    ConstSyst_ME1b = cms.double(0.02),
    ConstSyst_ME21 = cms.double(0.03),
    ConstSyst_ME22 = cms.double(0.03),
    ConstSyst_ME31 = cms.double(0.03),
    ConstSyst_ME32 = cms.double(0.03),
    ConstSyst_ME41 = cms.double(0.03),
    NoiseLevel_ME12 = cms.double(7.0),
    NoiseLevel_ME13 = cms.double(4.0),
    NoiseLevel_ME1a = cms.double(9.0),
    NoiseLevel_ME1b = cms.double(6.0),
    NoiseLevel_ME21 = cms.double(5.0),
    NoiseLevel_ME22 = cms.double(7.0),
    NoiseLevel_ME31 = cms.double(5.0),
    NoiseLevel_ME32 = cms.double(7.0),
    NoiseLevel_ME41 = cms.double(5.0),
    XTasymmetry_ME12 = cms.double(0.015),
    XTasymmetry_ME13 = cms.double(0.02),
    XTasymmetry_ME1a = cms.double(0.023),
    XTasymmetry_ME1b = cms.double(0.01),
    XTasymmetry_ME21 = cms.double(0.023),
    XTasymmetry_ME22 = cms.double(0.023),
    XTasymmetry_ME31 = cms.double(0.023),
    XTasymmetry_ME32 = cms.double(0.023),
    XTasymmetry_ME41 = cms.double(0.023)
)
```

**Comment by perrotta** [comment]
@jeongeun I supposed that we can discuss it by voice during our chat later on.

However, I think that the problem comes from the fact that the `cscRecHitDParameters` in `RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cff.py` are organized into a PSet, while in the code they are not supposed to be included in any PSet. As a consequence, the cloned config includes both the PSet and the same values automatically defined by the `fillDescriptions` method outside the PSet: hence the duplicate.

As of now I see two possible solutions (well, the third one being closing this PR, but I'd try to find a fix first):
- Remove the `RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cff.py` file and copy/paste all the values in the PSet, as well as all comments, inside `RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cfi.py`; the `cff` version is only used to be included in the `cfi`, and therefore there will be no other adjustments needed anywhere else
- Modiy the fillDescription code in the `CSCRecHitDProducer` so that those parameters are already organized into a PSet; this would allow retaining the current python structure

Perhaps @ptcox wants to add some comment or further suggestion....

**Comment by slava77** [comment]
> Modiy the fillDescription code in the `CSCRecHitDProducer` so that those parameters are already organized into a PSet; this would allow retaining the current python structure

I'd also suggest to rename `configWireTimeWindow` to `cscRecHitDProducer`, which is the standard default name

**Comment by perrotta** [comment]
@makortel @Dr15Jones perhaps this is something you can provide some suggestion (or fix, if needed).

In the config now cloned from the file automatically created by `CSCRecHitDProducer::fillDescriptions`, a set of parameters is imported from `cscRecHitDParameters`, where those parameters are organized in a `PSet`.
This results in a duplication of those imported parameters, as can be seen by symply expanding with  `edmConfigDump`:

```
$ edmConfigDump cscRecHitD_cfi.py
Traceback (most recent call last):
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/bin/slc7_amd64_gcc900/edmConfigDump", line 22, in <module>
    cfo = imp.load_source("pycfg", options.filename, handle)
  File "cscRecHitD_cfi.py", line 69, in <module>
    CSCStripClusterSize = 3
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Mixins.py", line 429, in clone
    **myparams)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Modules.py", line 166, in __init__
    super(EDProducer,self).__init__(type_,*arg,**kargs)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Modules.py", line 138, in __init__
    super(_Module,self).__init__(type_,*arg,**kargs)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Mixins.py", line 390, in __init__
    super(_TypedParameterizable,self).__init__(*arg,**kargs)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Mixins.py", line 177, in __init__
    self.__setParameters(kargs)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Mixins.py", line 260, in __setParameters
    self.__addParameter(name, value)
  File "/cvmfs/cms-ib.cern.ch/week0/slc7_amd64_gcc900/cms/cmssw/CMSSW_11_3_X_2021-04-13-1100/python/FWCore/ParameterSet/Mixins.py", line 249, in __addParameter
    raise ValueError(message)
ValueError: Duplicate insert of member XTasymmetry_ME1b
The original parameters are:
cms.EDProducer("CSCRecHitDProducer",
    ConstSyst_ME12 = cms.double(0.02),
    ConstSyst_ME13 = cms.double(0.03),
    ConstSyst_ME1a = cms.double(0.01),
    ConstSyst_ME1b = cms.double(0.02),
    ConstSyst_ME21 = cms.double(0.03),
    ConstSyst_ME22 = cms.double(0.03),
    ConstSyst_ME31 = cms.double(0.03),
    ConstSyst_ME32 = cms.double(0.03),
    ConstSyst_ME41 = cms.double(0.03),
    NoiseLevel_ME12 = cms.double(7.0),
    NoiseLevel_ME13 = cms.double(4.0),
    NoiseLevel_ME1a = cms.double(9.0),
    NoiseLevel_ME1b = cms.double(6.0),
    NoiseLevel_ME21 = cms.double(5.0),
    NoiseLevel_ME22 = cms.double(7.0),
    NoiseLevel_ME31 = cms.double(5.0),
    NoiseLevel_ME32 = cms.double(7.0),
    NoiseLevel_ME41 = cms.double(5.0),
    XTasymmetry_ME12 = cms.double(0.015),
    XTasymmetry_ME13 = cms.double(0.02),
    XTasymmetry_ME1a = cms.double(0.023),
    XTasymmetry_ME1b = cms.double(0.01),
    XTasymmetry_ME21 = cms.double(0.023),
    XTasymmetry_ME22 = cms.double(0.023),
    XTasymmetry_ME31 = cms.double(0.023),
    XTasymmetry_ME32 = cms.double(0.023),
    XTasymmetry_ME41 = cms.double(0.023)
)
```

Can this be considered the expected behavior for importing the parameters, or Is there anything to be fixed in the python import itself? IOW, should we wait for a central fix in the python configuration scripts, or fix orselves either the configuration or the `fillDescriptions` code in the `CSCRecHitDProducer` (as I suggested in https://github.com/cms-sw/cmssw/pull/33354#issuecomment-819259592)?

**Comment by makortel** [comment]
@perrotta Do I understand the situation correctly that the question is effectively what
```py
edmodule.clone(pset)
```
should do when `pset` is a `cms.PSet` that contains parameters already defined in `edmodule`? (that currently raises an exception on duplicate parameters)

**Comment by perrotta** [comment]
> @perrotta Do I understand the situation correctly that the question is effectively what
> 
> ```python
> edmodule.clone(pset)
> ```
> 
> should do when `pset` is a `cms.PSet` that contains parameters already defined in `edmodule`? (that currently raises an exception on duplicate parameters)

I'd say so, see https://github.com/cms-sw/cmssw/blob/f57494cb4aa1bed7f9b27eb69180f0ce9bad5e29/RecoLocalMuon/CSCRecHitD/python/cscRecHitD_cfi.py#L15

**Comment by perrotta** [comment]
@makortel , just to keep moving: is there any fix being prepared by the core team for the issue, or any suggestion about how to proceed? We can wait for it, if so. Otherwise, I'd suggest @jeongeun to proceed along either one of the two lines proposed in https://github.com/cms-sw/cmssw/pull/33354#issuecomment-819259592

**Comment by makortel** [comment]
@perrotta Apologies for delay. After some thought we concluded that `edmodule.clone(pset)` should be made to work, with the semantics that the values in `pset` override those of `edmodule`. Implementing that could take a few days of wall clock time.

One alternative would be to express the `cscRecHitDParameters` as a `dict` instead of `PSet`, and then use it in the clone along
```py
cscRecHitDParameters = dict(
    XTasymmetry_ME1a = 0.023,
    ...
)

csc2DRecHits = _mod.configWireTimeWindow.clone(
   ...
    **cscRecHitDParameters # must be last
)
```

**Comment by perrotta** [comment]
> @perrotta Apologies for delay. After some thought we concluded that `edmodule.clone(pset)` should be made to work, with the semantics that the values in `pset` override those of `edmodule`. Implementing that could take a few days of wall clock time.

Thank you Matti !

> One alternative would be to express the `cscRecHitDParameters` as a `dict` instead of `PSet`, and then use it in the clone along
> 
> ```python
> cscRecHitDParameters = dict(
>     XTasymmetry_ME1a = 0.023,
>     ...
> )
> 
> csc2DRecHits = _mod.configWireTimeWindow.clone(
>    ...
>     **cscRecHitDParameters # must be last
> )
> ```

Good point! This is less invasive than what I suggested [above](https://github.com/cms-sw/cmssw/pull/33354#issuecomment-819259592)

@jeongeun would you like to implement the suggestion of @makortel, and modify RecoLocalMuon/​CSCRecHitD/​python/​cscRecHitD_cff.py accordingly?

At the same time, I'd also profit to implement the [suggestion](https://github.com/cms-sw/cmssw/pull/33354#issuecomment-819448856) of @slava77 and rename the default filename `configWireTimeWindow` to `cscRecHitDProducer` in the fillDescriptions method of  the `CSCRecHitDProducer`

**Comment by jeongeun** [comment]
Sorry, I will close this PR and make another new PR for this modifications.
Thanks.

**Pull request closed without merging**